libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace std;
+using namespace solidity::frontend;
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	FunctionCallGraphBuilder builder;
+
+	builder.m_contract = &_contract;
+
+	builder.m_graph = make_unique<ContractCallGraph>(_contract);
+
+	// Create graph for constructor, state vars, etc
+	builder.m_currentNode = SpecialNode::EntryCreation;
+	builder.m_currentDispatch = SpecialNode::InternalCreationDispatch;
```

Comment:
These two could entirely be done in member-initializations, if you want (``m_currentDispatch`` already is initialized there - if you do it for ``m_currentNode`` as well, both can be removed here).

Reply:
I am not so sure it's good to do that though. The specific values they are initialized to very much relate to the actions done here. If we move this away and hide it in the header or constructor it might not be obvious why we set them to different values a few lines lower..

Reply:
Ok!